ground/src/main/java/com/google/android/ground/persistence/sync/LocalMutationSyncWorker.kt

```diff
@@ -80,63 +84,55 @@ constructor(
   private suspend fun getIncompleteMutations(): List<Mutation> =
     mutationRepository.getMutations(locationOfInterestId, PENDING, FAILED, IN_PROGRESS)
 
-  /**
-   * Groups mutations by user id, loads each user, applies mutations, and removes processed
-   * mutations.
-   *
-   * @return `true` if all mutations are applied successfully, else `false`
-   */
-  private suspend fun processMutations(allMutations: List<Mutation>): Boolean {
-    val mutationsByUserId = allMutations.groupBy { it.userId }
-    val userIds = mutationsByUserId.keys
-    var noErrors = true
-    for (userId in userIds) {
-      val mutations = mutationsByUserId[userId]
-      val user = getUser(userId)
-      if (mutations == null || user == null) {
-        continue
-      }
-      val result = processMutations(mutations, user)
-      if (!result) {
-        noErrors = false
-      }
-    }
-    return noErrors
-  }
-
   /**
    * Applies mutations to remote data store. Once successful, removes them from the local db.
    *
-   * @return `true` if the mutations were successfully synced with [RemoteDataStore].
+   * @return `success` if the mutations were successfully synced with [RemoteDataStore].
    */
-  private suspend fun processMutations(mutations: List<Mutation>, user: User): Boolean {
-    check(mutations.isNotEmpty()) { "List of mutations is empty" }
-
+  private suspend fun processMutations(mutations: List<Mutation>): Result {
+    if (mutations.isEmpty()) return success()
+    val user = getUserFromMutations(mutations) ?: return failure()
     return try {
       mutationRepository.markAsInProgress(mutations)
       remoteDataStore.applyMutations(mutations, user)
       mutationRepository.finalizePendingMutationsForMediaUpload(mutations)
-      true
+      success()
     } catch (t: Throwable) {
-      // Mark all mutations as having failed since the remote datastore only commits when all
-      // mutations have succeeded.
-      Timber.d(t, "Local mutation sync failed")
-      mutationRepository.markAsFailed(mutations, t)
-      val crashlytics = FirebaseCrashLogger()
-      crashlytics.setSelectedSurveyId(mutations.first().surveyId)
-      crashlytics.logException(t)
-      false
+      Timber.e(t)
+      handleMutationSyncFailed(mutations, t)
+      retry()
     }
   }
 
-  private suspend fun getUser(userId: String): User? {
-    val user = localUserStore.getUserOrNull(userId)
+  /**
+   * Returns the user associated with the mutations. If missing, removes the mutations from local
+   * storage as well.
+   */
+  private suspend fun getUserFromMutations(mutations: List<Mutation>): User? {
+    val userIds = mutations.map { it.userId }.toSet()
+
+    check(userIds.size == 1) { "Found mutations for multiple users" }
+    check(userRepository.getAuthenticatedUser().id == userIds.first()) {
```

Comment:
Let's `Timber.e()` and skip these instead to make sure data collectors don't end up with data stuck on their devices?

Reply:
Done